sample/sample.go

```diff
@@ -0,0 +1,192 @@
+package sample
+
+import (
+	"cmp"
+	"errors"
+	"math"
+	"slices"
+
+	pq "github.com/emirpasic/gods/v2/queues/priorityqueue"
+	"golang.org/x/exp/rand"
+	"gonum.org/v1/gonum/floats"
+	"gonum.org/v1/gonum/stat/sampleuv"
+)
+
+type Transform interface {
+	Apply([]float64) ([]float64, error)
+}
+
+type Sampler interface {
+	Sample([]float32, ...Transform) (int, error)
+}
+
+// TODO(parthsareen): potentially cache softmax values
+func softmax(logits []float64) []float64 {
+	copiedLogits := make([]float64, len(logits))
+	copy(copiedLogits, logits)
+	for i := range copiedLogits {
+		copiedLogits[i] = math.Exp(copiedLogits[i])
+	}
+
+	floatSum := floats.Sum(copiedLogits)
+	floats.Scale(1.0/floatSum, copiedLogits)
```

Comment:
You don't need an explicit copy. Instead sum and copy the values while iterating
```suggestion
	var sum float64
	tt := make([]float64, len(t))
	for i, v := range t {
		sum += math.Exp(v)
		tt[i] = math.Exp(v)
	}

	floats.Scale(1/sum, tt)
```

Reply:
Ah nice, will add